lib/uri.ml

```diff
@@ -707,11 +692,11 @@ let with_scheme uri =
   |Some scheme -> { uri with scheme=Some (Pct.cast_decoded scheme) }
   |None -> { uri with scheme=None }
 
-let host uri = get_decoded_opt uri.host
-let with_host uri =
-  function
-  |Some host -> { uri with host=Some (Pct.cast_decoded host) }
-  |None -> { uri with host=None }
+let host uri =
+  match uri.host with
+  | None -> None
+  | Some (`Ipv4_literal h | `Ipv6_literal h) -> Some h
```

Comment:
It's not clear to me whether the ipv6 here should be returned surrounded by square brackets. @mseri thinks it should, it would fix an issue in cohttp: https://github.com/mirage/ocaml-cohttp/issues/997

Reply:
It seems correct, the domain part of email addresses also use brackets to delimit user-defined domains or IPv6 address.

Reply:
Yes, as psafont mentioned above, @mseri proposed that it should be fixed here in the uri package in his comment https://github.com/mirage/ocaml-cohttp/issues/997#issuecomment-1721364362:
> I think this should be fixed in the uri package, then all dependent libraries would benefit directly from the fix.

@anmonteiro, can you have a look and update this PR to fix mirage/ocaml-cohttp#997?

Hier is the link to the code location in the full review diff:
https://github.com/mirage/ocaml-uri/pull/169/files#r1330212509

I don't know Ocaml, but a few lines above it, there is similar code (just with `host` instead of `h`), so maybe that can be used as a starting point for updating this function:
```ml
  |Some (`Ipv4_literal host) -> Buffer.add_string buf host
  |Some (`Ipv6_literal host) ->
      Buffer.add_char buf '[';
      Buffer.add_string buf host;
      Buffer.add_char buf ']'
  );
```